File: crates/agent-gateway/internal/stt/aliyun_dashscope.go
package stt

import (
	"context"
	"crypto/rand"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"strings"

	"github.com/gorilla/websocket"
)

type AliyunDashScopeAdapter struct{}

const dashScopeEndpoint = "wss://dashscope.aliyuncs.com/api-ws/v1/inference/"

func dashScopeModel(configured string) string {
	model := strings.TrimSpace(configured)
	if model == "" || model == "paraformer-realtime-8k-v2" {
		return "paraformer-realtime-v2"
	}
	return model
}

func dashScopeFinish(taskID string) map[string]any {
	return map[string]any{
		"header":  map[string]any{"action": "finish-task", "task_id": taskID, "streaming": "duplex"},
		"payload": map[string]any{"input": map[string]any{}},
	}
}

func (a *AliyunDashScopeAdapter) Test(ctx context.Context, cfg map[string]any) (string, error) {
	return readyProtocolTest(ctx, a, cfg)
}
func (a *AliyunDashScopeAdapter) Run(ctx context.Context, id string, cfg map[string]any, commands <-chan Command, events chan<- Event) error {
	endpoint, err := websocketEndpoint(cfg, dashScopeEndpoint)
	if err != nil {
		return stageError("DashScope", "validate", err)
	}
	model := dashScopeModel(value(cfg, "model"))
	var taskBytes [16]byte
	if _, err := rand.Read(taskBytes[:]); err != nil {
		return stageError("DashScope", "start", err)
	}
	wireTaskID := fmt.Sprintf("%x", taskBytes)
	header := http.Header{}
	header.Set("Authorization", "Bearer "+value(cfg, "apiKey"))
	header.Set("X-DashScope-DataInspection", "enable")
	conn, response, err := websocket.DefaultDialer.DialContext(ctx, endpoint, header)
	if err != nil {
		return stageError("DashScope", "connect", websocketConnectError(response, err))
	}
	defer func() { _ = conn.Close() }()
	start := map[string]any{
		"header": map[string]any{"action": "run-task", "task_id": wireTaskID, "streaming": "duplex"},
		"payload": map[string]any{
			"task_group": "audio", "task": "asr", "function": "recognition",
			"model": model,
			"parameters": map[string]any{
				"format": "pcm", "sample_rate": 16000,
				"language_hints":             []string{"zh", "en"},
				"max_sentence_silence":       2000,
				"disfluency_removal_enabled": false,
			},
			"input": map[string]any{},
		},
	}
	if err = writeProviderJSON(conn, start); err != nil {
		return stageError("DashScope", "start", err)
	}
	incoming := make(chan map[string]any, 8)
	readErr := make(chan error, 1)
	go func() {
		for {
			var msg map[string]any
			if err := conn.ReadJSON(&msg); err != nil {
				readErr <- stageError("DashScope", "receive", err)
				return
			}
			if !emitIncoming(ctx, incoming, msg) {
				return
			}
		}
	}()
	finishing := false
	finishSent := false
	ready := false
	pending := make([][]byte, 0, 32)
	finals := ""
	for {
		select {
		case <-ctx.Done():
			return nil
		case err := <-readErr:
			if finishing && finishSent && isWebSocketCloseError(err, websocket.CloseNormalClosure, websocket.CloseGoingAway) {
				return stageError("DashScope", "close", errors.New("connection closed before task-finished"))
			}
			return err
		case cmd := <-commands:
			if cmd.Cancel {
				return nil
			}
			if cmd.Audio != nil {
				if !ready {
					pending = append(pending, append([]byte(nil), cmd.Audio.PCM...))
				} else if err := writeProviderMessage(conn, websocket.BinaryMessage, cmd.Audio.PCM); err != nil {
					return stageError("DashScope", "send_audio", err)
				}
			}
			if cmd.Finish {
				finishing = true
				if ready && !finishSent {
					if err := writeProviderJSON(conn, dashScopeFinish(wireTaskID)); err != nil {
						return stageError("DashScope", "finish", err)
					}
					finishSent = true
				}
			}
		case msg := <-incoming:
			body, _ := json.Marshal(msg)
			var wire struct {
				Header struct {
					Event        string `json:"event"`
					ErrorCode    string `json:"error_code"`
					ErrorMessage string `json:"error_message"`
				} `json:"header"`
				Payload struct {
					Output struct {
						Sentence struct {
							Text        string `json:"text"`
							End         bool   `json:"end"`
							SentenceEnd bool   `json:"sentence_end"`
						} `json:"sentence"`
					} `json:"output"`
				} `json:"payload"`
			}
			if err := json.Unmarshal(body, &wire); err != nil {
				return stageError("DashScope", "parse", err)
			}
			switch wire.Header.Event {
			case "task-started":
				if !ready {
					ready = true
					if !emitEvent(ctx, events, Event{Type: "ready", SessionID: id}) {
						return ctx.Err()
					}
					for _, pcm := range pending {
						if err := writeProviderMessage(conn, websocket.BinaryMessage, pcm); err != nil {
							return stageError("DashScope", "send_audio", err)
						}
					}
					pending = nil
					if finishing && !finishSent {
						if err := writeProviderJSON(conn, dashScopeFinish(wireTaskID)); err != nil {
							return stageError("DashScope", "finish", err)
						}
						finishSent = true
					}
				}
			case "result-generated":
				if wire.Payload.Output.Sentence.End || wire.Payload.Output.Sentence.SentenceEnd {
					finals += wire.Payload.Output.Sentence.Text
					if !emitEvent(ctx, events, Event{Type: "partial", SessionID: id, Text: finals}) {
						return ctx.Err()
					}
				} else {
					if !emitEvent(ctx, events, Event{Type: "partial", SessionID: id, Text: finals + wire.Payload.Output.Sentence.Text}) {
						return ctx.Err()
					}
				}
			case "task-finished":
				if finishing && finishSent {
					if finals != "" {
						if !emitEvent(ctx, events, Event{Type: "final", SessionID: id, Text: finals}) {
							return ctx.Err()
						}
					}
					return nil
				}
			case "task-failed":
				return stageError("DashScope", "provider_response", providerFailure("DashScope", wire.Header.ErrorCode, wire.Header.ErrorMessage))
			}
		}
	}
}

File: crates/agent-gui/src-tauri/src/services/stt/aliyun_dashscope.rs
use super::{
    close_provider_socket, emit, provider_failure, send_provider_message, stage_failure, text,
    websocket_endpoint, SttCommand, SttEvent,
};
use futures_util::StreamExt;
use serde_json::Value;
use tauri::{AppHandle, Runtime};
use tokio::sync::mpsc::Receiver;
use tokio_tungstenite::{
    connect_async,
    tungstenite::{client::IntoClientRequest, http::header, Message},
};

fn dashscope_model(configured: String) -> String {
    if configured.is_empty() || configured == "paraformer-realtime-8k-v2" {
        "paraformer-realtime-v2".to_string()
    } else {
        configured
    }
}

fn finish_message(task_id: &str) -> Value {
    serde_json::json!({
        "header":{"action":"finish-task","task_id":task_id,"streaming":"duplex"},
        "payload":{"input":{}}
    })
}

pub async fn run<R: Runtime>(
    app: AppHandle<R>,
    session: String,
    config: serde_json::Map<String, Value>,
    mut rx: Receiver<SttCommand>,
) -> Result<(), String> {
    let key = text(&config, "apiKey");
    let endpoint = websocket_endpoint(&config, "wss://dashscope.aliyuncs.com/api-ws/v1/inference/")
        .map_err(|e| stage_failure("DashScope", "validate", e))?;
    let model = dashscope_model(text(&config, "model"));
    let task_id = uuid::Uuid::new_v4().simple().to_string();
    let mut request = endpoint
        .into_client_request()
        .map_err(|e| stage_failure("DashScope", "connect", e.to_string()))?;
    request.headers_mut().insert(
        header::AUTHORIZATION,
        format!("Bearer {key}")
            .parse()
            .map_err(|_| "无效 API Key")?,
    );
    request.headers_mut().insert(
        "X-DashScope-DataInspection",
        "enable"
            .parse()
            .map_err(|_| "DashScope 数据检查请求头无效")?,
    );
    let (socket, _) = connect_async(request)
        .await
        .map_err(|e| stage_failure("DashScope", "connect", format!("网络错误: {e}")))?;
    let (mut write, mut read) = socket.split();
    send_provider_message(&mut write, Message::Text( serde_json::json!({ "header":{"action":"run-task","task_id":task_id,"streaming":"duplex"}, "payload":{ "task_group":"audio","task":"asr","function":"recognition", "model":&model, "parameters":{ "format":"pcm", "sample_rate":16000, "language_hints":["zh", "en"], "max_sentence_silence":2000, "disfluency_removal_enabled":false }, "input":{} } }) .to_string() .into(), ), "DashScope", "start").await?;
    let mut ready = false;
    let mut finishing = false;
    let mut finish_sent = false;
    let mut pending = Vec::<Vec<u8>>::new();
    let mut finals = String::new();
    loop {
        tokio::select! {
            Some(command) = rx.recv() => match command {
                SttCommand::Audio { pcm, .. } => { if ready { send_provider_message(&mut write, Message::Binary(pcm.into()), "DashScope", "send_audio").await?; } else { pending.push(pcm); } }
                SttCommand::Finish => {
                    if !finishing { finishing = true; if ready { send_provider_message(&mut write, Message::Text(finish_message(&task_id).to_string().into()), "DashScope", "finish").await?; finish_sent = true; } }
                }
                SttCommand::Cancel => { close_provider_socket(&mut write).await; return Ok(()); }
            },
            Some(message) = read.next() => {
                let message = message.map_err(|e| stage_failure("DashScope", "receive", e.to_string()))?;
                let Message::Text(body) = message else { continue };
                let value: Value = serde_json::from_str(&body).map_err(|_| stage_failure("DashScope", "parse", "返回内容不是有效 JSON"))?;
                match value.pointer("/header/event").and_then(Value::as_str).unwrap_or_default() {
                    "task-started" => {
                        if !ready {
                            ready = true;
                            emit(&app, SttEvent::Ready { session_id: session.clone() });
                            for pcm in pending.drain(..) { send_provider_message(&mut write, Message::Binary(pcm.into()), "DashScope", "send_audio").await?; }
                            if finishing && !finish_sent {
                                send_provider_message(&mut write, Message::Text(finish_message(&task_id).to_string().into()), "DashScope", "finish").await?;
                                finish_sent = true;
                            }
                        }
                    }
                    "result-generated" => {
                        let sentence = value.pointer("/payload/output/sentence");
                        let sentence_text = sentence.and_then(|s| s.get("text")).and_then(Value::as_str).unwrap_or_default();
                        let ended = sentence.and_then(|s| s.get("end")).and_then(Value::as_bool).unwrap_or(false) || sentence.and_then(|s| s.get("sentence_end")).and_then(Value::as_bool).unwrap_or(false);
                        if ended { finals.push_str(sentence_text); }
                        let partial = if ended { finals.clone() } else { format!("{finals}{sentence_text}") };
                        emit(&app, SttEvent::Partial { session_id: session.clone(), text: partial });
                    }
                    "task-finished" if finish_sent => {
                        if !finals.is_empty() { emit(&app, SttEvent::Final { session_id: session.clone(), text: finals.clone() }); }
                        close_provider_socket(&mut write).await;
                        return Ok(());
                    }
                    "task-failed" => return Err(stage_failure("DashScope", "provider_response", provider_failure(
                        "DashScope",
                        value.pointer("/header/error_code").and_then(Value::as_str).unwrap_or_default(),
                        value.pointer("/header/error_message").and_then(Value::as_str).unwrap_or_default(),
                    ))),
                    _ => {}
                }
            }
            else => return Err(stage_failure("DashScope", "close", "连接在 task-finished 前关闭")),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn uses_16_khz_model_and_complete_finish_message() {
        assert_eq!(dashscope_model(String::new()), "paraformer-realtime-v2");
        assert_eq!(
            dashscope_model("paraformer-realtime-8k-v2".to_string()),
            "paraformer-realtime-v2"
        );
        let finish = finish_message("task");
        assert_eq!(
            finish.pointer("/header/streaming").and_then(Value::as_str),
            Some("duplex")
        );
        assert!(finish.pointer("/payload/input").is_some());
    }
}
